### ISSUE TITLE
PuObjectRequest causes weird error with docker

### ISSUE THREAD

**Issue by vicmosin** [findify/s3mock#83]
Once I run the mock as a docker container it fails on 
```
PutObjectRequest putObjectRequest = new PutObjectRequest(ensureCreateBucket(bucketId), fileId, file)
       .withMetadata(meta);
storage.putObject(putObjectRequest);
```
with

```
com.amazonaws.services.s3.model.AmazonS3Exception:  (Service: Amazon S3; Status Code: 500; Error Code: InternalError; Request ID: null; S3 Extended Request ID: null)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1639) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1304) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1056) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667) ~[aws-java-sdk-core-1.11.243.jar:na]
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649) ~[aws-java-sdk-core-1.11.243.jar:na]
```

Docker logs whows the following:
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.iq80.leveldb.util.ByteBufferSupport (file:/app/s3mock.jar) to method java.nio.DirectByteBuffer.cleaner()
WARNING: Please consider reporting this to the maintainers of org.iq80.leveldb.util.ByteBufferSupport
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
``` 

InMemory tests from java go through. Any ideas?

**Comment by bateo102** [comment]
I am getting the same error although it seems that the objects do get placed in the bucket.  S3 seems unable to return PutObjectResult object.

**Comment by liufuyang** [comment]
The same issue encountered using docker image on mac Bucket creation works, upload file does not.

**Comment by max8github** [comment]
Please review patch 5ddf3c8 as it contains a workaround to this issue.
By publishing a docker image with a tag similar to `0.2.4-jdk-8-slim` in dockerhub, people can use a functioning docker container (the current 0.2.4 does not work - should be removed).

(Incidentally, on your main [dockerhub](https://hub.docker.com/r/findify/s3mock) page, could you please correct this line:`docker exec -p 8001:8001 findify/s3mock:latest` to this:
```
docker run -p 8001:8001 findify/s3mock:latest
```
)

**Comment by max8github** [comment]
Here is the issue:
```
ByteBufferSupport cannot access class jdk.internal.ref.Cleaner (in module java.base) because module java.base does not export jdk.internal.ref to unnamed module @373e8ef5
```
The above happens when a `putObject` is issued. In FileProvider, the file gets saved, but not the metadata. The call `metadataStore.put(bucket, key, objectMetadata)` will cause an exception that will leave a file locked (a `LOCK` file in the metadata directory for the bucket).
The offending code is in library leveldb, class  `DbLock`. In its constructor, tryLock throws the exception above leaving metadata in an inconsistent state, thus compromising any successive `getObject` call (getObject will fail as it will find a lock).

All of that only happens with java 9, not with java 8. 
See https://bugs.openjdk.java.net/browse/JDK-8148117.

So for all that, the easy workaround is to either use java 8, or to use S3Mock's in memory provider (`InMemoryProvider`) instead of `FileProvider`.

Going to create a docker image tagging it with java 9 and "inmem".

**Comment by max8github** [comment]
Verified that with `leveldb` verison `0.10` this issue is resolved.

**Comment by slouc** [comment]
@max8github Using your in-memory image prevents the exception, but I always seem to read 0 bytes in the object content. Object itself is found correctly, but `getObjectContent` returns an empty `InputStream`. It is also possible that the content is not being written properly, although inspecting the object before writing shows proper content with properly set content length and other metadata, so as far as I can debug, it looks OK. But either it doesn't really write properly, or there's something weird upon reading. I'm using the same code that works with fake-s3 (I wanted to use a Scala library instead of a Ruby one).

Did anyone else perhaps encounter this? I can try to recreate a minimal example, but first I wanted to check if someone already had this problem and figured it out.

**Comment by ppurang** [comment]
An addendum to @slouc 's comment. 

We tried @max8github  's image with his fixes. But ran into the [signed API issue](https://github.com/findify/s3mock/issues/88) so we worked around it using this 

```scala
          .withCredentials(new AWSStaticCredentialsProvider(new AnonymousAWSCredentials()))
```

**Comment by max8github** [comment]
I have also found other problems. One of them has to do with `DatatypeConverter` not being available in the regular java 9 bundle:
```
java.lang.NoClassDefFoundError: javax/xml/bind/DatatypeConverter
at better.files.File.checksum(File.scala:395)
at better.files.File.md5(File.scala:398)
at io.findify.s3mock.provider.FileProvider.putObjectMultipartComplete(FileProvider.scala:119)
at io.findify.s3mock.route.PutObjectMultipartComplete.$anonfun$route$5(PutObjectMultipartComplete.scala:22)
at scala.util.Try$.apply(Try.scala:209)
at io.findify.s3mock.route.PutObjectMultipartComplete.$anonfun$route$4(PutObjectMultipartComplete.scala:22)
at akka.http.scaladsl.server.directives.RouteDirectives.$anonfun$complete$1(RouteDirectives.scala:47)
```
So for that and other issues, i created an image based on java 8 instead (after applying the same fix, that is upgrading leveldb to 0.10). I would suggest using that one (0.2.4.1-8-jdk-slim) instead of the java 9 one. See https://hub.docker.com/r/max8github/s3mock/tags/

**Comment by chrisandrews7** [comment]
When will you be publishing the fix to docker hub?
